source/common/network/buffered_io_socket_handle_impl.h

```diff
@@ -0,0 +1,145 @@
+#pragma once
+
+#include <memory>
+
+#include "envoy/api/io_error.h"
+#include "envoy/api/os_sys_calls.h"
+#include "envoy/common/platform.h"
+#include "envoy/event/dispatcher.h"
+#include "envoy/network/io_handle.h"
+
+#include "common/buffer/watermark_buffer.h"
+#include "common/common/logger.h"
+#include "common/event/user_space_file_event_impl.h"
+#include "common/network/io_socket_error_impl.h"
+#include "common/network/peer_buffer.h"
+
+namespace Envoy {
+namespace Network {
+
+/**
+ * IoHandle implementation which provides a buffer as data source. It is designed to used by
+ * Network::ConnectionImpl. Some known limitations include
+ * 1. It doesn't not include a file descriptor. Do not use "fdDoNotUse".
+ * 2. It doesn't support socket options. Wrap this in ConnectionSocket and implement the socket
+ * getter/setter options.
+ * 3. It doesn't support UDP interface.
+ * 4. The peer BufferedIoSocket must be scheduled in the same thread to avoid data race because
+ *    BufferedIoSocketHandle mutates the state of peer handle and no lock is introduced.
+ */
+class BufferedIoSocketHandleImpl : public IoHandle,
+                                   public ReadWritable,
+                                   protected Logger::Loggable<Logger::Id::io> {
+public:
+  BufferedIoSocketHandleImpl();
+
+  ~BufferedIoSocketHandleImpl() override { ASSERT(closed_); }
```

Comment:
This is not a valid ASSERT.  IoSocketHandleImpl calls close from it's destructor.

Reply:
That is huge break! 
Will fix